src/coreclr/jit/objectalloc.cpp

```diff
@@ -1334,3 +1482,1724 @@ void ObjectAllocator::RewriteUses()
+{
+    JITDUMP("Checking if [%06u] in " FMT_BB " executes under a %s GDV type test\n", comp->dspTreeID(tree), block->bbNum,
+            testOutcome ? "successful" : "failing");
+
+    // Walk up the dominator tree....
+    //
+    for (BasicBlock* idomBlock = block->bbIDom; idomBlock != nullptr; idomBlock = idomBlock->bbIDom)
+    {
+        JITDUMP("... examining dominator " FMT_BB "\n", idomBlock->bbNum);
+        if (!idomBlock->KindIs(BBJ_COND))
+        {
+            JITDUMP("... not BBJ_COND\n");
+            continue;
+        }
+
+        // We require that one idomBlock successor *not* dominate.
+        // (otherwise idomBlock could be the top of a diamond where both outcomes reach block).
+        //
+        const bool trueSuccessorDominates  = comp->m_domTree->Dominates(idomBlock->GetTrueTarget(), block);
+        const bool falseSuccessorDominates = comp->m_domTree->Dominates(idomBlock->GetFalseTarget(), block);
+
+        if (trueSuccessorDominates && falseSuccessorDominates)
+        {
+            JITDUMP("... both successors dominate?\n");
+            continue;
+        }
+
+        if (!(trueSuccessorDominates || falseSuccessorDominates))
+        {
+            JITDUMP("... neither successor dominates\n");
+            continue;
+        }
+
+        GenTree* const guardingRelop = IsGuard(idomBlock, info);
+        if (guardingRelop == nullptr)
+        {
+            continue;
+        }
+
+        // We found a dominating GDV test, see if the condition is the one we're looking for.
+        //
+        if (testOutcome)
+        {
+            bool const isReachableOnGDVSuccess = (trueSuccessorDominates && guardingRelop->OperIs(GT_EQ)) ||
+                                                 (falseSuccessorDominates && guardingRelop->OperIs(GT_NE));
+            if (isReachableOnGDVSuccess)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by failing GDV\n");
+            continue;
+        }
+        else
+        {
+            bool const isReachableOnGDVFailure = (trueSuccessorDominates && guardingRelop->OperIs(GT_NE)) ||
+                                                 (falseSuccessorDominates && guardingRelop->OperIs(GT_EQ));
+            if (isReachableOnGDVFailure)
+            {
+                info->m_block = idomBlock;
+                return true;
+            }
+            JITDUMP("... guarded by successful GDV\n");
+            continue;
+        }
+    }
+
+    JITDUMP("... no more doms\n");
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// IsGuard: does block look like a GDV guard
+//
+// Arguments:
+//   block -- block in question
+//   info -- [out] guard info
+//
+// Returns:
+//   Comparison tree if this is a guard, or nullptr
+//
+GenTree* ObjectAllocator::IsGuard(BasicBlock* block, GuardInfo* info)
+{
```

Comment:
Can this function be used for GDV chaining as well? (to remove the `GTF_CALL_M_GUARDED_DEVIRT_CHAIN` flag)

Reply:
Possibly? I haven't looked.